### DISCUSSION TITLE
The problem of obtaining the BSDF at the intersection

### DISCUSSION THREAD

**Discussion by ioissss** [mitsuba-renderer/mitsuba3#1205] [Q&A] [answered]
I'm writing my own Integrator, inheriting from mitsuba.SamplingIntegrator, rewriting the sample(self,scene,sampler,ray,medium=None,active=True) function in it, and my sample() function is as follows:

```py
def sample(self,scene,sampler,ray,medium=None,active=True):
        L = mi.Color3f(0.0) #initialize radiance
        si = scene.ray_intersect(ray,active=active) #intersection
        active &= si.is_valid()
        # NEE
        ctx = mi.BSDFContext()
        ctx.type_mask = mi.BSDFFlags.All
        ctx.mode = mi.TransportMode.Radiance
        bsdf = si.bsdf()
        emitter_sample,emitter_weight = scene.sample_emitter_direction(si,sampler.next_2d(),test_visibility=True,active=active)
        data = np.array(bsdf.eval(ctx,si,emitter_sample.d))
        np.savetxt('./output.txt',data,fmt="%.3f")
        #if si is a emitter
        L += si.emitter(scene).eval(si)
        L += emitter_weight*bsdf.eval(ctx,si,emitter_sample.d)
        return [L,active,[]] 
```
Running this code, I get the following results:
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/91249397/4e77fb1a-b357-44f6-86b4-8ec6377b6fc5)
There's a big part that's black.

I checked the running result of ’bsdf.eval(ctx,si,emitter_sample.d)‘ and found that most of them are (0.0, 0.0, 0.0),  print(bsdf) can be found that there are many of them are None.
 the scenario that I used is the tutorial cbox.xml.
I didn't find any examples related to rewriting Integrator in the tutorials, where can I find such examples, the API documentation describes it very briefly, where can I find examples of usage of related functions

**Comment by ioissss** [comment]
I visualized the intersection of bsdf.eval() value, found that the problem appeared in the bsdf value of the evaluation, each intersection at the evaluation of the bsdf value after visualization is as follows: 
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/91249397/f76099dc-a165-4472-8f84-e3de6b56f541)
I don't know why I'm getting this error, is there something wrong with the emitter_sample.d ? I've always assumed that emitter_sample.d is pointing from the intersection point to the light source and is located in the world coordinate system, the code is 

```py
bsdf = si.bsdf()
emitter_sample,emitter_weight = scene.sample_emitter_direction(si,sampler.next_2d(),test_visibility=True,active=active)

L += emitter_weight*bsdf.eval(mi.BSDFContext(),si,emitter_sample.d)
```

**Comment by njroussel** [comment]
When opening an issue please follow the pre-filled template :warning: 

I'll move this to a discussion thread for now, it seems more appropriate.

**Comment by njroussel** [comment, marked as answer]
Hi @ioissss 

For examples of integrators you can check out the source code of all the integrators we ship. You'll find it [here](https://github.com/mitsuba-renderer/mitsuba3/tree/master/src/python/python/ad/integrators) and [here](https://github.com/mitsuba-renderer/mitsuba3/tree/master/src/integrators).

As for your error, I believe that the issue is that the outgoing direction `emitter_sample.d` is in world-space, when it should be in local-space. You can fix it with the following:
```python
wo = si.to_local(emitter_sample.d);
L += emitter_weight*bsdf.eval(mi.BSDFContext(), wo)
```

**Reply by ioissss** [reply]
thank you for your reply! when using the function bsdf.eval(), the wo should be in local space and i incorrectly provided the wo in world space
